### PR TITLE
MAINT: Start applying ruff rules (RUF)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Apply additional rules in preparation for https://github.com/numpy/numpy/issues/24994.

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -3387,13 +3387,13 @@ class ufunc:
...
     # arguments and return one output argument. Because of that we
     # can't type them very precisely.
-    def reduce(self, /, *args: Any, **kwargs: Any) -> NoReturn | Any: ...
-    def accumulate(self, /, *args: Any, **kwargs: Any) -> NoReturn | NDArray[Any]: ...
-    def reduceat(self, /, *args: Any, **kwargs: Any) -> NoReturn | NDArray[Any]: ...
-    def outer(self, *args: Any, **kwargs: Any) -> NoReturn | Any: ...
+    def reduce(self, /, *args: Any, **kwargs: Any) -> Any: ...
+    def accumulate(self, /, *args: Any, **kwargs: Any) -> NDArray[Any]: ...
+    def reduceat(self, /, *args: Any, **kwargs: Any) -> NDArray[Any]: ...
+    def outer(self, *args: Any, **kwargs: Any) -> Any: ...
```
Comment (charris) [inline numpy/__init__.pyi:3393]: Hmm, NoReturn seems odd here, makes me wonder why it was there.

Comment (charris) [reply numpy/__init__.pyi:3393]: @jorenham Thoughts?

**Review by charris — APPROVED** [review]
LGTM, but I have left a question.

File: numpy/__init__.pyi (hunk 1 above)
Comment (jorenham) [reply numpy/__init__.pyi:3393]: `NoReturn | T` is equivalent to `T`, as `NoReturn` is the "bottom type" (just like `Never`). You removing it makes sense 👌🏻

File: numpy/lib/_nanfunctions_impl.py [hunk 2]
```diff
@@ -1662,7 +1663,7 @@ def _nanquantile_ureduce_func(
         a: np.array,
         q: np.array,
         weights: np.array,
-        axis: int = None,
+        axis: Optional[int] = None,
```
Comment (jorenham) [inline numpy/lib/_nanfunctions_impl.py:1666]:
```suggestion
        axis: int | None = None,
```

**Review by jorenham — CHANGES_REQUESTED** [review]

File: numpy/f2py/capi_maps.py [hunk 3]
```diff
@@ -627,7 +627,7 @@ def routsign2map(rout):
                                 ln = k
                                 break
                     lcb_map[ln] = un[1]
-    elif 'externals' in rout and rout['externals']:
+    elif rout.get('externals'):
```
[the remark this answers is not in the document]
Comment (DimitriPapadopoulos) [reply numpy/f2py/capi_maps.py:630]: I'm not sure about that:
- before: `'externals' in rout` returns `True`, but then `rout['externals']` returns `None`
- after: `rout.get('externals')` returns `None`

The initial code tests that `rout` contains an `'externals'` value that is not `None`.

File: numpy/core/_utils.py [hunk 4]
```diff
@@ -1,7 +1,8 @@
 import warnings
+from typing import Optional
 
 
-def _raise_warning(attr: str, submodule: str = None) -> None:
+def _raise_warning(attr: str, submodule: Optional[str] = None) -> None:
```
[the remark this answers is not in the document]
Comment (DimitriPapadopoulos) [reply numpy/core/_utils.py:5]: I cannot commit this suggestion. Is it a diff or a suggestion?

File: numpy/f2py/capi_maps.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorenham) [reply numpy/f2py/capi_maps.py:630]: oh yea nevermind, I missed the part after the `and ...` for some reason 😅

File: numpy/lib/_nanfunctions_impl.py [hunk 5]
```diff
@@ -27,6 +27,7 @@
 from numpy.lib import _function_base_impl as fnb
 from numpy.lib._function_base_impl import _weights_are_valid
 from numpy._core import overrides
+from typing import Optional
```
Comment (charris) [inline numpy/lib/_nanfunctions_impl.py:30]: Is this import still needed?

File: numpy/lib/_function_base_impl.py [hunk 6]
```diff
@@ -30,6 +30,7 @@
 
 # needed in this module for compatibility
 from numpy.lib._histograms_impl import histogram, histogramdd  # noqa: F401
+from typing import Optional
```
Comment (charris) [inline numpy/lib/_function_base_impl.py:33]: Is this import still needed?

File: numpy/lib/_nanfunctions_impl.py (hunk 5 above)
Comment (DimitriPapadopoulos) [reply numpy/lib/_nanfunctions_impl.py]: Fixed.

**Review by jorenham — APPROVED** [review]

**Review by mhvk — APPROVED** [review]
All nice changes!

**Pull request merged**